File: 111.minimum-depth-of-binary-tree.java
/*
 * @lc app=leetcode id=111 lang=java
 *
 * [111] Minimum Depth of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
;
        return lengthOfTree(root, 1);
    }

    public int lengthOfTree(TreeNode root, int currentLength) {
        if (root.left == null && root.right == null) {
            return currentLength;
        }
        int minLeft=Integer.MAX_VALUE,minRight=Integer.MAX_VALUE;
        if (root.left != null) {

         minLeft = lengthOfTree(root.left, currentLength + 1);
        }
        if (root.right != null) {

         minRight = lengthOfTree(root.right, currentLength + 1);
        }
        return Math.min(minLeft, minRight);
    }
}
// @lc code=end

File: 111.minimum-depth-of-binary-tree.py
#
# @lc app=leetcode id=111 lang=python3
#
# [111] Minimum Depth of Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from typing import Optional


class Solution:
    
    def dfs(self, root: TreeNode, depth: int) -> int:
        if depth > self.global_min:
            return self.global_min
        if root.left is None and root.right is None:
            self.global_min = min(self.global_min, depth + 1)
            return depth + 1
        
        left,right = 1000001,1000001
        if root.left :
            left = self.dfs(root.left, depth + 1)
        if root.right :
            right = self.dfs(root.right, depth + 1)
        return min(left, right)
    
    def minDepth(self, root: Optional[TreeNode]) -> int:
        self.global_min = 1000001
        return 0 if root is None else self.dfs(root, 0)
        

        
# @lc code=end

